pack_service.cmake

```diff
@@ -0,0 +1,252 @@
+    json_append_comma(result)
+    json_append_string(result ${KEY})
+    string(APPEND result ":")
+    string(APPEND result "[")
+    foreach(item IN LISTS ARGN)
+        json_append_comma(result)
+        json_append_string(result ${item})
+    endforeach()
+    string(APPEND result "]")
+    set(${VAR} ${result} PARENT_SCOPE)
+endfunction()
+
+function(write_meta)
+    cmake_parse_arguments(PARSE_ARGV 0 "" "" "NAME;DESCRIPTION;OUTPUT" "DEPENDS;ACCOUNTS")
+    set(result)
+    string(APPEND result "{")
+    json_append_key(result "name" ${_NAME})
+    json_append_key(result "description" ${_DESCRIPTION})
+    json_append_list(result "depends" ${_DEPENDS})
+    json_append_list(result "accounts" ${_ACCOUNTS})
+    string(APPEND result "}")
+    file(GENERATE OUTPUT ${_OUTPUT} CONTENT ${result})
+endfunction()
+
+function(write_service_info)
+    cmake_parse_arguments(PARSE_ARGV 0 "" "" "SERVER;OUTPUT" "INIT;FLAGS")
+    set(result)
+    string(APPEND result "{")
+    json_append_list(result "flags" ${_FLAGS})
+    if (_SERVER)
+        json_append_key(result "server" ${_SERVER})
+    endif()
+    if (_INIT)
+        json_append_comma(result)
+        json_append_string(result "init")
+        string(APPEND result ":")
+        string(APPEND result true)
+    endif()
+    string(APPEND result "}")
+    file(GENERATE OUTPUT ${_OUTPUT} CONTENT ${result})
+endfunction()
+
+# NAME <name>               - The name of the package
+# DESCRIPTION <text>        - The package description
+# OUTPUT <filename>         - The package file. Defaults to ${NAME}.psi
+# ACCONTS <name>...         - Additional non-service accounts to create
+# DEPENDS <targets>...      - Targets that this target depends on
+# PACKAGE_DEPENDS <name>... - Other psibase packages that the package depends on
+# SERVICE <name>            - The account name of a service
+#   WASM <filename>         - The wasm file that will be deployed
+#   FLAGS <flags>...        - Flags to set on the service account (requires chain admin powers to install)
+#   SERVER <name>           - The service that handles HTTP requests
+#   DATA <path> <dest>      - Uploads a file or directory to the target location
+#   DATA GLOB <path>... <dir> - Uploads multiple files to a directory
+#   INIT                    - The service has an init action that should be run with no arguments
+function(psibase_package)
+    set(keywords NAME DESCRIPTION OUTPUT PACKAGE_DEPENDS DEPENDS ACCOUNTS SERVICE DATA WASM FLAGS SERVER INIT POSTINSTALL)
+    foreach(keyword IN LISTS keywords)
+        set(_${keyword})
+    endforeach()
+    set(_SERVICES)
+    math(EXPR end ${ARGC}-1)
+    foreach(n RANGE ${end}})
+        set(my_keyword)
+        foreach(keyword IN LISTS keywords)
+            if (ARGV${n} STREQUAL keyword)
+                set(my_keyword ${keyword})
+            endif()
+        endforeach()
+        if(my_keyword)
+            if(current_keyword STREQUAL "DATA")
+                list(APPEND _DATA_${_SERVICE} DATA)
```

Comment:
Is this meant to append the string "DATA"?

Reply:
Yes